// gitbook-plugin-theme-appcom/src/js/theme/sidebar.js
var $ = require('jquery');

var platform = require('./platform');

var gitbook = window.gitbook;

// Toggle sidebar with or withour animation
function toggleSidebar(_state, animation) {
    if (gitbook.state != null && isOpen() == _state) 
        return;
    if (animation == null) 
        animation = true;
    
    gitbook
        .state
        .$book
        .toggleClass('without-animation', !animation);
    gitbook
        .state
        .$book
        .toggleClass('with-summary', _state);

    gitbook
        .storage
        .set('sidebar', isOpen());
}

// Return true if sidebar is open
function isOpen() {
    return gitbook
        .state
        .$book
        .hasClass('with-summary');
}

// Prepare sidebar: state and toggle button
function init() {
    // Init last state if not mobile
    if (!platform.isMobile()) {
        toggleSidebar(gitbook.storage.get('sidebar', true), false);
    }

    // Close sidebar after clicking a link on mobile
    $(document)
        .on('click', '.book-summary li.chapter a', function (e) {
            if (platform.isMobile()) 
                toggleSidebar(false, false);
            }
        );
}

// Filter summary with a list of path
function filterSummary(paths) {
    var $summary = $('.book-summary');

    $summary
        .find('li')
        .each(function () {
            var path = $(this).data('path');
            var st = paths == null || paths.indexOf(path) !== -1;

            $(this).toggle(st);
            if (st) 
                $(this).parents('li').show();
            }
        );
}

function expand(chapter) {
    chapter.show();

    if (chapter.parent().attr('class') != 'summary' && chapter.parent().attr('class') != 'book-summary' && chapter.length != 0) {
        expand(chapter.parent());
    }
}

gitbook
    .events
    .bind("page.change", function () {
        $('li.chapter')
            .children('ul.articles')
            .hide();
        $chapter = $('li.chapter.active');
        $children = $chapter.children('ul.articles');
        $parent = $chapter.parent();
        $siblings = $chapter
            .siblings()
            .children('ul.articles');

        expand($chapter);

        if ($children.length > 0) {
            $children.show();
        }
    });

module.exports = {
    init: init,
    isOpen: isOpen,
    toggle: toggleSidebar,
    filter: filterSummary
};
